src/Lean/Elab/Deriving/Dynamic.lean

```diff
@@ -0,0 +1,26 @@
+/-
+Copyright (c) 2022 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Gabriel Ebner
+-/
+import Lean.Elab.Deriving.Basic
+import Std.Any
+
+namespace Lean.Elab
+open Command Std Parser Term
+
+private def deriveDynamicInstance (declNames : Array Name) : CommandElabM Bool := do
+  for declName in declNames do
+    let cinfo ← getConstInfo declName
+    unless cinfo.levelParams.isEmpty do
+      throwError m!"{mkConst declName} has universe level parameters"
+    elabCommand <| ← withFreshMacroScope `(
+      unsafe def instImpl : Dynamic @$(mkCIdent declName) := .mk _ $(quote declName)
+      @[implementedBy instImpl] opaque inst : Dynamic @$(mkCIdent declName)
+      instance : Dynamic @$(mkCIdent declName) := inst
```

Comment:
We can't mark it unsafe because the instance needs to be available in safe definitions.

What would be possible is `@[implementedBy instImpl] opaque inst : ...`, but I wanted to get the default instance name.